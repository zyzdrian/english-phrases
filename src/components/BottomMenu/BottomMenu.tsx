import React from 'react';
import styled from 'styled-components';
import { Button, IconButton } from '..';

export enum BOTTOM_MENU_TEST_IDS {
    bottomMenu = 'bottom-menu',
    undoButton = 'undo-button',
    showMeaningButton = 'show-meaning-button',
    iKnowButton = 'i-know-button',
    iDontKnowButton = 'i-dont-know-button'
}

const StyledBottomMenu = styled.div`
display: flex;
padding: 24px;
max-width: 600px;
width: 100%;

& button {
  margin-right: 8px;
  
  &:last-child {
    margin-right: 0;
  }
}
`;

export const BottomMenu = ({ markAsKnown, markAsNotKnown, rollbackState, enableButtons, areButtonsDisabled, isRollbackDisabled }: IBottomMenuProps) => (
    <StyledBottomMenu data-testid={BOTTOM_MENU_TEST_IDS.bottomMenu}>
        {areButtonsDisabled ? (
            <>
                <IconButton variant='secondary' onClick={rollbackState} disabled={isRollbackDisabled} data-testid={BOTTOM_MENU_TEST_IDS.undoButton} />
                <Button onClick={enableButtons} data-testid={BOTTOM_MENU_TEST_IDS.showMeaningButton}>SHOW MEANING</Button>
            </>
        ): (
            <>
                <Button variant='secondary' onClick={markAsNotKnown} disabled={areButtonsDisabled} data-testid={BOTTOM_MENU_TEST_IDS.iDontKnowButton}>
                    {'I don\'t know'}
                </Button>
                <Button onClick={markAsKnown} disabled={areButtonsDisabled} data-testid={BOTTOM_MENU_TEST_IDS.iKnowButton}>I know</Button>
            </>
        )}
    </StyledBottomMenu>
);

export interface IBottomMenuProps {
    markAsKnown: () => void;
    markAsNotKnown: () => void;
    rollbackState: () => void;
    enableButtons: () => void;
    isRollbackDisabled: boolean;
    areButtonsDisabled: boolean;
}
