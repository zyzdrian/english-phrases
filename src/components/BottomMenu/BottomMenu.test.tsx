import React from 'react';
import { BottomMenu, IBottomMenuProps, BOTTOM_MENU_TEST_IDS } from './BottomMenu';
import { cleanup, render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

describe('<BottomMenu /> tests', () => {
    const INITIAL_PROPS: IBottomMenuProps = {
        areButtonsDisabled: false,
        markAsKnown: jest.fn(),
        markAsNotKnown: jest.fn(),
        rollbackState: jest.fn(),
        enableButtons: jest.fn(),
        isRollbackDisabled: false
    };
    let props = { ...INITIAL_PROPS };
    beforeEach(() => {
        props = { ...INITIAL_PROPS };
    });
    afterEach(cleanup);

    it('renders without disabled props', () => {
        const { container, getByTestId } = render(<BottomMenu {...props} />);
        const expectedChildrenNumber = 2;

        expect(container.firstChild).toMatchSnapshot();
        expect(getByTestId(BOTTOM_MENU_TEST_IDS.iKnowButton)).toBeTruthy();
        expect(getByTestId(BOTTOM_MENU_TEST_IDS.iDontKnowButton)).toBeTruthy();
        expect(getByTestId(BOTTOM_MENU_TEST_IDS.bottomMenu).children.length).toEqual(expectedChildrenNumber);
    });

    it('renders clickable "I know" button', () => {
        const { container, getByTestId } = render(<BottomMenu {...props} />);
        fireEvent.click(getByTestId(BOTTOM_MENU_TEST_IDS.iKnowButton));

        expect(container.firstChild).toMatchSnapshot();
        expect(props.markAsKnown).toBeCalled();
    });

    it('renders clickable "I don\'t know" button', () => {
        const { container, getByTestId } = render(<BottomMenu {...props} />);
        fireEvent.click(getByTestId(BOTTOM_MENU_TEST_IDS.iDontKnowButton));

        expect(container.firstChild).toMatchSnapshot();
        expect(props.markAsKnown).toBeCalled();
    });

    it('renders other buttons when areButtonsDisabled = true', () => {
        props.areButtonsDisabled = true;

        const { container, getByTestId } = render(<BottomMenu {...props}/>);

        expect(container.firstChild).toMatchSnapshot();
        expect(getByTestId(BOTTOM_MENU_TEST_IDS.showMeaningButton)).toBeTruthy();
        expect(getByTestId(BOTTOM_MENU_TEST_IDS.undoButton)).toBeTruthy();
        expect(getByTestId(BOTTOM_MENU_TEST_IDS.undoButton)).not.toBeDisabled();
    });

    it('renders other buttons when areButtonsDisabled = true and isRollbackDisabled = true', () => {
        props.areButtonsDisabled = true;
        props.isRollbackDisabled = true;

        const { container, getByTestId } = render(<BottomMenu {...props}/>);

        expect(container.firstChild).toMatchSnapshot();
        expect(getByTestId(BOTTOM_MENU_TEST_IDS.showMeaningButton)).toBeTruthy();
        expect(getByTestId(BOTTOM_MENU_TEST_IDS.undoButton)).toBeTruthy();
        expect(getByTestId(BOTTOM_MENU_TEST_IDS.undoButton)).toBeDisabled();
    });

    it('renders clickable "undo button"', () => {
        props.areButtonsDisabled = true;

        const { getByTestId } = render(<BottomMenu {...props}/>);
        fireEvent.click(getByTestId(BOTTOM_MENU_TEST_IDS.undoButton));

        expect(props.rollbackState).toBeCalled();
    });

    it('renders clickable "Show meaning button"', () => {
        props.areButtonsDisabled = true;

        const { getByTestId } = render(<BottomMenu {...props}/>);
        fireEvent.click(getByTestId(BOTTOM_MENU_TEST_IDS.showMeaningButton));

        expect(props.enableButtons).toBeCalled();
    });
});
