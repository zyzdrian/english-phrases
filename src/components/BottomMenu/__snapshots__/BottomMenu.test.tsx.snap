// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BottomMenu /> tests renders clickable "I don't know" button 1`] = `
<div
  class="sc-cSHVUG fwdZJS"
  data-testid="bottom-menu"
>
  <button
    class="sc-bdVaJa fnRXyy"
    data-testid="i-dont-know-button"
  >
    I don't know
  </button>
  <button
    class="sc-bdVaJa fnRXyy"
    data-testid="i-know-button"
  >
    I know
  </button>
</div>
`;

exports[`<BottomMenu /> tests renders clickable "I know" button 1`] = `
<div
  class="sc-cSHVUG fwdZJS"
  data-testid="bottom-menu"
>
  <button
    class="sc-bdVaJa fnRXyy"
    data-testid="i-dont-know-button"
  >
    I don't know
  </button>
  <button
    class="sc-bdVaJa fnRXyy"
    data-testid="i-know-button"
  >
    I know
  </button>
</div>
`;

exports[`<BottomMenu /> tests renders other buttons when areButtonsDisabled = true 1`] = `
<div
  class="sc-cSHVUG fwdZJS"
  data-testid="bottom-menu"
>
  <button
    class="sc-bdVaJa fnRXyy sc-bZQynM igNFdC"
    data-testid="undo-button"
  >
    <img
      alt="<-"
      src="left-arrow.svg"
    />
  </button>
  <button
    class="sc-bdVaJa fnRXyy"
    data-testid="show-meaning-button"
  >
    SHOW MEANING
  </button>
</div>
`;

exports[`<BottomMenu /> tests renders other buttons when areButtonsDisabled = true and isRollbackDisabled = true 1`] = `
<div
  class="sc-cSHVUG fwdZJS"
  data-testid="bottom-menu"
>
  <button
    class="sc-bdVaJa fnRXyy sc-bZQynM igNFdC"
    data-testid="undo-button"
    disabled=""
  >
    <img
      alt="<-"
      src="left-arrow.svg"
    />
  </button>
  <button
    class="sc-bdVaJa fnRXyy"
    data-testid="show-meaning-button"
  >
    SHOW MEANING
  </button>
</div>
`;

exports[`<BottomMenu /> tests renders without disabled props 1`] = `
<div
  class="sc-cSHVUG fwdZJS"
  data-testid="bottom-menu"
>
  <button
    class="sc-bdVaJa fnRXyy"
    data-testid="i-dont-know-button"
  >
    I don't know
  </button>
  <button
    class="sc-bdVaJa fnRXyy"
    data-testid="i-know-button"
  >
    I know
  </button>
</div>
`;
